core/src/main/scala/kafka/coordinator/group/GroupCoordinator.scala

```diff
@@ -118,23 +118,30 @@ class GroupCoordinator(val brokerId: Int,
       sessionTimeoutMs > groupConfig.groupMaxSessionTimeoutMs) {
       responseCallback(joinError(memberId, Errors.INVALID_SESSION_TIMEOUT))
     } else {
+      val isUnknownMember = memberId == JoinGroupRequest.UNKNOWN_MEMBER_ID
       groupManager.getGroup(groupId) match {
         case None =>
           // only try to create the group if the group is UNKNOWN AND
           // the member id is UNKNOWN, if member is specified but group does not
           // exist we should reject the request.
-          if (memberId == JoinGroupRequest.UNKNOWN_MEMBER_ID) {
+          if (isUnknownMember) {
             val group = groupManager.addGroup(new GroupMetadata(groupId, Empty, time))
             doUnknownJoinGroup(group, requireKnownMemberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
           } else {
             responseCallback(joinError(memberId, Errors.UNKNOWN_MEMBER_ID))
           }
 
         case Some(group) =>
-          if (memberId == JoinGroupRequest.UNKNOWN_MEMBER_ID) {
-            doUnknownJoinGroup(group, requireKnownMemberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
-          } else {
-            doJoinGroup(group, memberId, clientId, clientHost, rebalanceTimeoutMs, sessionTimeoutMs, protocolType, protocols, responseCallback)
+          group.inLock {
+            if (groupIsOverCapacity(group) || (isUnknownMember && groupIsFull(group))) {
```

Comment:
I think we just need to kick out real members until the `groupIsFull` is met. Probably we could just use groupIsFull to keep rejecting new members, while doesn't care about pending members to make the handling logic simpler.

Reply:
I don't think that would work.
If we were to detect that `groupIsOverCapacity()` and start kicking out members until `groupIsFull`, we might kick out members *before* they had a chance to commit offsets.

When we receive a`JoinGroup` request from a member, we know that it has attempted to commit offsets. (see `ConsumerCoordinator#onJoinPrepare`)